### PR TITLE
Buzzer API

### LINKED ISSUE #1: Add the Buzzer API and music example
Based on the LED API present here, implement the Buzzer API and port the music example from libtock-c. You can use a micro:bit v2 as hardware.

Documentation:
- libtock-rs overview - https://github.com/tock/libtock-rs/blob/master/doc/Overview.md
- Tock system calls - https://github.com/tock/tock/blob/master/doc/Syscalls.md
- LED API - https://github.com/tock/libtock-rs/tree/master/apis/leds
- The music example - https://github.com/tock/libtock-c/tree/master/examples/music
- microbit:v2 - https://microbit.org/new-microbit/

### PR DESCRIPTION
**Pull request by SheepSeb**
## Pull Request Overview

This pull request continues the work done in #4 by @teodorobert

This pull request adds:
- a buzzer API resolves #1 , similar to the [libtock-c version]( https://github.com/tock/libtock-c/blob/master/libtock/buzzer.h ). The API provides a function that simulates a synchronous tone setting and also functions to register/unregister a listener, check for driver existence
- an example app which plays Ode of Joy
- a ```fake``` buzzer driver, for testing
- unit tests for the fake driver and for the user space library

## Testing Strategy
This PR was tested with the unit tests and by running the example app on a microbit v2 board

## TODO or Help Wanted
Feedback

## Documentation Updated
- [x] no updates are required

### CODE REVIEW

File: apis/buzzer/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+impl<S: Syscalls> Buzzer<S> {
+    /// Returns Ok() if the driver was present.This does not necessarily mean
+    /// that the driver is working.
+    pub fn exists() -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, EXISTS, 0, 0).to_result()
+    }
+
+    /// Initiate a tone
+    pub fn tone(freq: u32, duration: u32) -> Result<(), ErrorCode> {
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs:20]:
```suggestion
    pub fn tone(freq: u32, duration: Duration) -> Result<(), ErrorCode> {
```

Comment (alexandruradovici) [reply apis/buzzer/src/lib.rs:20]: Use the [`core::time::Duration`](https://doc.rust-lang.org/core/time/struct.Duration.html) data type.

File: apis/buzzer/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+        S::command(DRIVER_NUM, BUZZER_ON, freq, duration).to_result()
+    }
+
+    /// Register an events listener
+    pub fn register_listener<'share, F: Fn(i32)>(
+        listener: &'share BuzzerListener<F>,
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs:26]:
```suggestion
        listener: &'share Cell<(u32,)><F>,
```

File: apis/buzzer/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+        });
+
+        match buzzer_cell.get() {
+            None => Err(ErrorCode::Busy),
+            Some(_buzzer_val) => Ok(()),
+        }
+    }
+}
+
+pub struct BuzzerListener<F: Fn(i32)>(pub F);
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs:61]: This is not needed.

File: examples/music.rs [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+    for (frequency, duration) in MELODY.iter() {
+        let mut note_duration = WHOLE_NOTE / duration.unsigned_abs();
+        if duration < &0 {
+            note_duration = note_duration * 15 / 10;
+        }
+
+        let note_duration = note_duration * 9 / 10;
+        Buzzer::tone_sync(frequency * 3, note_duration).unwrap();
+    }
```
Comment (alexandruradovici) [inline examples/music.rs:101]: Move this in the `Ok(_)` branch.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Comment by alexandruradovici** [comment]
@SheepSeb any updates?

File: apis/buzzer/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    /// Unregister the events listener
+    pub fn unregister_listener() {
+        S::unsubscribe(DRIVER_NUM, 0)
+    }
+
+    /// Initiate a synchronous tone
+    /// Returns Ok() if the operation was successful
+    pub fn tone_sync(freq: u32, duration: Duration) -> Result<u32, ErrorCode> {
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs:38]:
```suggestion
    pub fn tone_sync(freq: u32, duration: Duration) -> Result<(), ErrorCode> {
```

File: apis/buzzer/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+                if let Ok(()) = Self::tone(freq, duration) {
+                    while listener.get() == None {
+                        S::yield_wait();
+                    }
+                }
+            }
+        });
+
+        match listener.get() {
+            None => Err(ErrorCode::Busy),
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs:51]: Why does this mean that the buzzer if busy? I think it means failure.

File: apis/buzzer/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            Some(buzzer_val) => Ok(buzzer_val.0),
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs:52]:
```suggestion
            Some(0) => Ok(()),
            Some(buzzer_val) => Err(ErrorCode::try_from_u32(buzzer_val.0).map_or(|| ErrorCode::BadRVal))
```

something similar

File: apis/buzzer/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    }
+
+    /// Initiate a tone
+    pub fn tone(freq: u32, duration: Duration) -> Result<(), ErrorCode> {
+        S::command(DRIVER_NUM, BUZZER_ON, freq, duration.as_millis() as u32).to_result()
+    }
+
+    /// Register an events listener
+    pub fn register_listener<'share>(
+        listener: &'share Cell<Option<(u32,)>>,
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs:25]: I would add a listener, as the listener should not return an u32, rather an Result<(), ErrorCode>

File: apis/buzzer/src/lib.rs [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+
+const DRIVER_NUM: u32 = 0x90000;
+
+// Command IDs
+const EXISTS: u32 = 0;
+const BUZZER_ON: u32 = 1;
+
+/// The notes that can be played by the buzzer
+#[allow(unused)]
+pub mod note {
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs:72]:
```suggestion
#[repr(u32))
pub enum Note {
```

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

File: apis/buzzer/src/lib.rs (hunk 8 above)
Comment (alexandruradovici) [reply apis/buzzer/src/lib.rs:25]: I think we did this wrong here. We need to be able to register a listener that the user can actually use outside this crate, meaning that the listener needs to have a function that is called when the upcall is fired. 

Something similar to the [`ButtonListener`](https://github.com/tock/libtock-rs/blob/cff868b1fb2641f2d1b0f3edaae827cbcfdadef3/apis/buttons/src/lib.rs#L124) should work.

**Review by alexandruradovici — CHANGES_REQUESTED** [review]

**Review by alexandruradovici — DISMISSED** [review]
Looks good, send it to upstream.

File: apis/buzzer/src/lib.rs (hunk 9 above)
Comment (alexandruradovici) [reply apis/buzzer/src/lib.rs]: Fix this.

File: apis/buzzer/src/lib.rs [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+        let listener = Cell::new(Some((0,)));
+        share::scope(|subscribe| {
+            if let Ok(()) = Self::register_listener(&listener, subscribe) {
+                if let Ok(()) = Self::tone(freq, duration) {
+                    while listener.get() == None {
+                        S::yield_wait();
+                    }
+                }
+            }
+        });
```
Comment (alexandruradovici) [inline apis/buzzer/src/lib.rs]: Use ?

Comment (SheepSeb) [reply apis/buzzer/src/lib.rs]: This code is outdated in my branch is like [this](https://github.com/SheepSeb/libtock-rs/blob/sebnae/buzzer/apis/buzzer/src/lib.rs)

**Comment by SheepSeb** [comment]
Merged into master

**Pull request closed without merging**